tsdb/tsm1/reader_range_maxtime_iterator.go

```diff
@@ -0,0 +1,154 @@
+package tsm1
+
+import (
+	"math"
+)
+
+const (
+	// InvalidMinNanoTime is an invalid nano timestamp that has an ordinal
+	// value lower than models.MinNanoTime.
+	InvalidMinNanoTime = math.MinInt64
+)
+
+// TimeRangeMaxTimeIterator will iterate over the keys of a TSM file, starting at
+// the provided key. It is used to determine if each key has data which exists
+// within a specified time interval.
+type TimeRangeMaxTimeIterator struct {
+	timeRangeBlockReader
+
+	// cached values
+	lastWrite int64
+	hasData   bool
+	isLoaded  bool
+}
+
+// Next advances the iterator and reports if it is still valid.
+func (b *TimeRangeMaxTimeIterator) Next() bool {
+	if b.Err() != nil {
+		return false
+	}
+
+	b.clearIsLoaded()
+
+	return b.iter.Next()
+}
+
+// Seek points the iterator at the smallest key greater than or equal to the
+// given key, returning true if it was an exact match. It returns false for
+// ok if the key does not exist.
+func (b *TimeRangeMaxTimeIterator) Seek(key []byte) (exact, ok bool) {
+	if b.Err() != nil {
+		return false, false
+	}
+
+	b.clearIsLoaded()
+
+	return b.iter.Seek(key)
+}
+
+// HasData reports true if the current key has data for the time range.
+func (b *TimeRangeMaxTimeIterator) HasData() bool {
+	if b.Err() != nil {
+		return false
+	}
+
+	b.load()
+
+	return b.hasData
+}
+
+// MaxTime returns the maximum timestamp for the current key within the
+// requested time rage. If an error occurred or there is no data,
+// InvalidMinTimeStamp will be returned, which is less than models.MinTimeStamp.
+// This property can be leveraged when enumerating keys to find the maximum timestamp,
+// as this value will always be lower than any valid timestamp returned.
+func (b *TimeRangeMaxTimeIterator) MaxTime() int64 {
+	if b.Err() != nil {
+		return InvalidMinNanoTime
+	}
+
+	b.load()
+
+	return b.lastWrite
+}
+
+func (b *TimeRangeMaxTimeIterator) clearIsLoaded() { b.isLoaded = false }
+
+func (b *TimeRangeMaxTimeIterator) load() {
+	if b.isLoaded {
+		return
+	}
+
+	b.isLoaded = true
+	b.hasData = false
+	b.lastWrite = InvalidMinNanoTime
+
+	e, ts := b.getEntriesAndTombstones()
+	if len(e) == 0 {
+		return
+	}
+
+	if len(ts) == 0 {
+		b.lastWrite = e[len(e)-1].MaxTime
+		// no tombstones, fast path will avoid decoding blocks
+		// if queried time interval intersects with one of the entries
+		if intersectsEntry(e, b.tr) {
+			b.hasData = true
+			return
+		}
+
+		for i := range e {
```

Comment:
I think I may have got the wrong end of the stick, but my assumption here is that these entries may or may not contain data within `b.tr`. And you have to decode the block to actually check for points that overlap `b.tr`?

I could think of one example of this being a block that has a time-range of `0–10` and the time-range you care about is `2–4`.

If my understanding of what is going on here is correct then I think you can improve the performance a bit here. Index entries should be stored in TSM files ordered by `min time`. I think the code that handles that is here:

https://github.com/influxdata/influxdb/blob/45250eeda2901cf9843cf63f9d0a9c4de626d23b/tsdb/tsm1/writer.go#L495-L498

Therefore, you should be able to exit early once an index entry's `min` time is greater than `b.tr.Max`.

Or maybe that's handled elsewhere?

Reply:
Agreed; the code now reflects that 👍